ghcide/src/Development/IDE/Core/ProgressReporting.hs

```diff
@@ -42,38 +48,60 @@ data ProgressEvent
   | ProgressCompleted
   | ProgressStarted
 
-data ProgressReporting m = ProgressReporting
-  { progressUpdate :: ProgressEvent -> m (),
-    inProgress     :: forall a. NormalizedFilePath -> m a -> m a,
-    -- ^ see Note [ProgressReporting API and InProgressState]
-    progressStop   :: IO ()
+data ProgressReportingNoTrace m = ProgressReportingNoTrace
+  { progressUpdateI :: ProgressEvent -> m (),
+    progressStopI   :: IO ()
     -- ^ we are using IO here because creating and stopping the `ProgressReporting`
     -- is different from how we use it.
   }
 
+data ProgressReporting m = ProgressReporting
+  {
+    inProgress             :: forall a. NormalizedFilePath -> m a -> m a,
+    -- ^ see Note [ProgressReporting API and InProgressState]
+    progressReportingInner :: ProgressReportingNoTrace m
+  }
+
+class ProgressReportingClass a where
+    type M a :: * -> *
```

Comment:
I think I asked this before, but can everything just be in IO? The only place we use another monad is in `inProgress`, and I think we could also just provide that with IO and get people to use `MonadUnliftIO` to use it?

Reply:
`inProgress`'s monad `m` is invariant (Both covariant and contra variant locations). I doubt we can have `forall a.IO a -> IO a` -> `forall a. m a -> m a`.🤔

Reply:
Yes, we can turn them into IOs and save the `m`